---
title: "Computational workflow for the identification of Methylation QTLs (meQTL) in DNA methylation and genotyped datasets"
author: "Weixiong He"
date: "01/07/2018"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      message = TRUE,warning = FALSE)
```


```{r library, message=FALSE}
library(edgeR)
library(tidyverse)
library(magrittr)
library(devtools)
library(biomaRt)
library(MatrixEQTL)
library(limma)
library(devtools)
library(ggman)
library(readr)
library(lumi)
library(ggfortify)
library(rtracklayer)
library(FactoMineR)
library(plyranges)
library(GWASTools)
library(SNPRelate)
library(GenomicRanges)
library(genomation)
library(Gviz)
library(grange)
library(coMET)
library(GenomicInteractions)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
```


## Set working directory
```{r}
setwd("~/R/Weixiong_Project/data/Weixiong_DNAme_files")
```

## Preparing the genome annotation

Ensembl is a global resource for bioinformatics, and contains information on the annotation of eukaryotic genomes. Using the package `biomaRt` we will firstly obtain the snp and gene information for hg19/GRCh37


```{r dataset}
grch37_snp = useEnsembl(biomart="ENSEMBL_MART_SNP", 
                 host="grch37.ensembl.org", 
                 dataset = "hsapiens_snp")

grch37_gene = useEnsembl(biomart="ENSEMBL_MART_ENSEMBL", 
                 host="grch37.ensembl.org", 
                 dataset = "hsapiens_gene_ensembl")

```

## Gentoype data 

Genotype data are usually VCF format, we convert them to GDS to extract the genotype profile and SNP location.

```{r}
showfile.gds(closeall=TRUE)
vcf.fn <- "nihRnaEarlyGestation_mafFiltered_1kgFiltered_87Samples_rmDup.vcf.gz"
snpgdsVCF2GDS(vcf.fn, "genotypes.gds", method="biallelic.only", verbose = TRUE)
genofile <- snpgdsOpen("genotypes.gds")
snpgdsSummary(genofile)
snpset <- snpgdsLDpruning(genofile, ld.threshold=0.8, slide.max.n = 50, maf=0.1, remove.monosnp = TRUE)
snpset.id.pruned <- unlist(snpset)

g <- read.gdsn(index.gdsn(genofile, "genotype"))

g <- g[,snpset.id.pruned]

g <- g %>% t %>% as.data.frame()
sample_id <- (read.gdsn(index.gdsn(genofile, "sample.id")))

colnames(g) <- sample_id

ids <- (read.gdsn(index.gdsn(genofile, "snp.rs.id")))
idsPruned <- ids[snpset.id.pruned]
genotypes <- tibble::add_column(g, rsIDs = idsPruned, .before = 1) 
write_tsv(genotypes, "genotype_prunned_87_0.8_nodup.tsv")

## snp_position 

pos <- read.gdsn(index.gdsn(genofile, "snp.position"))
posPruned <- pos[snpset.id.pruned]
chrom <- read.gdsn(index.gdsn(genofile,"snp.chromosome"))
chromPruned <- chrom[snpset.id.pruned]
snp_posPruned <- cbind(rsIDs = idsPruned,chr = chromPruned,pos=posPruned) %>% as.data.frame()
write_tsv(snp_posPruned,"snp_position_pruned_0.8_nodup.tsv")
```

## Preparing methylation data 

We can either use beta- or M-value matirx. To display the difference between beta- and M-value, we generate density plot. 

```{r SNP sampleInfo}
#loading placental methylation data
normDNAme <- readRDS("Placenta_betaBMIQ_96.rds") 

DNAmeSampleInfo <- read_delim("DNAme_placenta_sampleInfo.txt",
                              delim = "\t",col_names = c("Name", "ID", "GestationalAge", "Slide", "Basename",
                                            "Gender", "MaternalAge", "BMI", "SmokingStatus", "Batch"),
                              col_types = "ccicccidcc") %>% as.data.frame()
normDNAme <- as.data.frame(normDNAme)
colnames(normDNAme) <- DNAmeSampleInfo$ID
pdf("Beta-value.pdf", w=11, h=7, pointsize=15)
densityPlot(as.matrix(normDNAme), sampGroups=DNAmeSampleInfo$ID,
            main="Density plot of Beta-value matrix", legend=FALSE)
dev.off()
pdf("M-value.pdf", w=11, h=7, pointsize=15)
densityPlot(as.matrix(normDNAme_M), sampGroups=colnames(normDNAme_M),legend=FALSE,main="Density plot of M-value matrix",xlab="M-values")
dev.off()
normDNAme_M <- beta2m(normDNAme)
normDNAme_87_beta <- normDNAme %>% 
  dplyr::select(colnames(genotypes[,2:length(colnames(genotypes))])) %>% 
  rownames_to_column(var="ID")
write_tsv(normDNAme_87_beta,"methy_int_87_beta.tsv")

##PCA
methylPCA_M <- normDNAme_87 %>% column_to_rownames(var="ID") %>% PCA()
PC1 <- methylPCA_M$var$coord[,1]
PC2 <- methylPCA_M$var$coord[,2]
pairs(methylPCA_M$var$coord)
labs <- rownames(methylPCA_M$var$coord)
PCs <- data.frame(cbind(PC1,PC2))
rownames(PCs) <- labs
df_out <- as.data.frame(methylPCA_M$ind$coord)

ggplot(PCs, aes(PC1,PC2, label=rownames(PCs))) +geom_text(aes(color=rownames(PCs)))+ scale_color_viridis(option="magma",discrete=TRUE)+theme_bw()+ theme(legend.position = "none") + xlab("PC1") + ylab("PC2") +
  coord_cartesian(xlim =c(.8,1), ylim = c(-.3, .3)) +
  ggsave("meth_PCA.pdf",dpi = 320,height=6,width = 6)

```

## Preparing position information of CpG probes

Get position information from EPIC Array manifest file

```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
ann.Epic <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19) #get annotation of methylation data
probes <- as.data.frame(ann.Epic) %>% #annotation data -> dataframe
  dplyr::select(chr, pos) %>% #extract chromsome and position information
  tibble::rownames_to_column(var = "id") %>% #id: col -> row
  mutate(chr = gsub("^chr", "", chr)) %>% #replace ^chr to chr
  mutate(s2 = pos+1,
         s1 = pos) %>% ## 1bp
  dplyr::select(id, chr, start=s1, end=s2)

write_tsv(probes,"probes.tsv")
```

## covariants data

Add covariants to count other variations in linear model

```{r PCA_5_covariates}
pca <- snpgdsPCA(genofile,snp.id=snpset.id.pruned)
pc.percent <- pca$varprop*100
head(round(pc.percent, 2))
tab <- data.frame(sample.id = pca$sample.id,
    EV1 = pca$eigenvect[,1],    # the first eigenvector
    EV2 = pca$eigenvect[,2],    # the second eigenvector
    stringsAsFactors = FALSE)
head(tab)

# lbls <- paste("PC", 1:4, "\n", format(pc.percent[1:4], digits=2), "%", sep="")
# pairs(pca$eigenvect[,1:4], labels=lbls)
ggplot(tab,aes(x=EV1,y=EV2,label =tab$sample.id,color=factor(sample.id))) + 
  geom_text(show.legend = FALSE) + 
  labs(x="PC1",y="PC2") + 
coord_cartesian(xlim =c(-1,1), ylim = c(-1, 1))+
  theme_bw() +
  ggsave("Genotype_PCA.pdf",dpi = 320,height=7,width = 7)

# Take out first 3 pc and add to covariates
pcaSamples <- t(pca$eigenvect[,1:3]) %>%
  as.data.frame()
colnames(pcaSamples) <- colnames(sample)
pcaCovs <- t(pcaSamples) %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ID")

colnames(pcaCovs) <- c("ID", "PCA1", "PCA2", "PCA3")


cov <- read_delim("covariates_87.txt","\t", escape_double = FALSE, trim_ws = TRUE) %>% column_to_rownames(var="ID") %>% 
  t() %>% 
  as.data.frame() %>% 
  dplyr::select(GestationalAge,Sex,RIN) %>% 
  rownames_to_column(var="ID") 
  
PEER_covariates <- read_delim("M2.PEER_covariates.txt", delim = "\t", col_names = FALSE) %>% 
  as.data.frame()
colnames(PEER_covariates) <- PEER_covariates[1,]
PEER_covariates <- PEER_covariates[-1,]
rownames(PEER_covariates) <- PEER_covariates[,1]
PEER_covariates <- PEER_covariates %>% t %>% as.data.frame() 
PEER_covariates <- PEER_covariates[-1,] %>% rownames_to_column(var="ID")
covariates <- inner_join(cov, pcaCovs) %>%
  inner_join(PEER_covariates) %>% 
  column_to_rownames(var="ID") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var="ID")

write_tsv(covariates,"covariates_87_new.tsv")

png("cov_sex.png", w=2000, h=1800, res = 200)

read_delim("covariates_87_new.tsv","\t", escape_double = FALSE, trim_ws = TRUE) %>% column_to_rownames(var="ID") %>% 
  t() %>% 
  as.data.frame() %>%  
  ggplot(aes(x=PCA1,y=PCA2,color = as.factor(Sex),label =cov$ID)) + 
  geom_text() + 
  labs(colour="Sex",title="Sex")+ 
  scale_color_manual(labels=c("male","female"),values = c("#0099CC","#FF6666"))+
  coord_cartesian(xlim =c(-2,2), ylim = c(-2, 2)) + 
  theme_bw()+theme(plot.title = element_text(hjust = 0.5)) 
dev.off()
png("cov_GA.png", w=2300, h=1800, res = 200)
read_delim("covariates_87_new.tsv","\t", escape_double = FALSE, trim_ws = TRUE) %>% column_to_rownames(var="ID") %>% 
  t() %>% 
  as.data.frame() %>%
ggplot(aes(x=PCA1,y=PCA2,colour = factor(GestationalAge),label =cov$ID)) + 
  geom_text(show.legend = FALSE) + 
  labs(title="Gestational Age")+
  coord_cartesian(xlim =c(-2,2), ylim = c(-2, 2))+theme_bw()+
  theme(plot.title = element_text(hjust = 0.5)) 
dev.off()
## RNA integrity number
png("cov_RIN.png", w=2300, h=1800, res = 200)
read_delim("covariates_87_new.tsv","\t", escape_double = FALSE, trim_ws = TRUE) %>% column_to_rownames(var="ID") %>% 
  t() %>% 
  as.data.frame() %>%
ggplot(aes(x=PCA1,y=PCA2,colour = RIN,label =cov$ID)) + 
  geom_text(show.legend = FALSE) + 
  labs(title="RNA integrity number")+
  coord_cartesian(xlim =c(-2,2), ylim = c(-2, 2))+
  theme_bw() +theme(plot.title = element_text(hjust = 0.5)) 
dev.off()


#plot(cov$PCA1,cov$PCA2,xlab="PCA1", ylab="PCA2", col=brewer.pal(n = 2, name = "Paired"))
```




## Run MatrixEQTL

```{r MatrixEQTL, message=FALSE}
snps <- SlicedData$new()
snps$fileDelimiter = "\t"   # the TAB character
snps$fileOmitCharacters = "NA" # denote missing values;
snps$fileSkipRows = 1        # one row of column labels
snps$fileSkipColumns = 1      # one column of row labels
snps$fileSliceSize = 10000      # read file in slices of 100000 rows
snps$LoadFile("genotype_prunned_87_0.8_nodup.tsv")

### Load gene expression info
DNAme = SlicedData$new()
DNAme$fileDelimiter = "\t"      # the TAB character
DNAme$fileOmitCharacters = "character" # denote missing values;
DNAme$fileSkipRows = 1        # one row of column labels
DNAme$fileSkipColumns = 1       # one column of row labels
DNAme$fileSliceSize = 10000     # read file in slices of 2,0000 rows
DNAme$LoadFile("methy_int_87_beta.tsv")

### Load covariates
cvrt = SlicedData$new()
cvrt$fileDelimiter = "\t"
cvrt$fileOmitCharacters = "NA"
cvrt$fileSkipRows = 1
cvrt$fileSkipColumns = 1
cvrt$LoadFile("covariates_87_new.tsv");

common = intersect(DNAme$columnNames, snps$columnNames) 

snps$ColumnSubsample(match(common, snps$columnNames)) 

DNAme$ColumnSubsample(match(common, DNAme$columnNames))
cvrt$ColumnSubsample(match(common, cvrt$columnNames)) #combin three datas

snpspos = read.table("snp_position_pruned_0.8_nodup.tsv", header = TRUE, stringsAsFactors = FALSE) #load 
DNAmepos = read.table("probes_f.tsv", header = TRUE, stringsAsFactors = FALSE)

options(MatrixEQTL.dont.preserve.gene.object = TRUE)

### Run Matrix eQTL
me_DNAme_beta = Matrix_eQTL_main(
  snps = snps,
  gene = DNAme,
  cvrt = cvrt,
  output_file_name = 'DNAme_output_trans_1e-8.txt', 
  pvOutputThreshold = 1e-8,  ###If the results of the associate study exceed the threshold, the results will not be retained in the final document
  useModel = modelLINEAR,
  errorCovariance = numeric(),
  verbose = TRUE,  # Setting verbose to True in the configuration will result in the service generating more output (will show both WARNING and INFO log levels), normally you will only see WARNING or higher (ERROR for example)
  output_file_name.cis = "DNAme_output_cis_1e-5.txt",
  pvOutputThreshold.cis = 1e-5,
  snpspos = snpspos,
  genepos = DNAmepos,
  # Distance for local gene-SNP pairs
  cisDist = 1e6,      #### 1e6 cis window = 1Mb (GTEx) ### DNAme 50kb 
  noFDRsaveMemory = FALSE, ##save significant gene-SNP pairs directly to the output files, reduce memory footprint and skip FDR calculation. The eQTLs are not recorded in the returned object if noFDRsaveMemory = TRUE
  min.pv.by.genesnp = TRUE, ## record the minimum p-value for each SNP and each gene in the returned object. The minimum p-values are recorded even if if they are above the corresponding thresholds of pvOutputThreshold and pvOutputThreshold.cis.
  pvalue.hist = "qqplot") ## record information sufficient to create a QQ- plot of the p-values 


save(me_DNAme_beta, file = "me_QTL_B_nopdup.Rdata")

```



# How many meQTLs were found

```{r}
me_DNAme_beta$all$neqtls
me_DNAme$cis$neqtls
me_DNAme$trans$neqtls
me_DNAme$time.in.sec
me_DNAme$all$ntests
```

## Manhattan Plot
```{r Manhattan Plot}
## Manhathan plot
library(qqman)
pvalue_cis <- as.data.frame(me_DNAme$cis$min.pv.snps) %>% 
  tibble::rownames_to_column(var = "ID")

pvalue_cis

colnames(pvalue_cis) <- c("ID","pvalue")
head(pvalue_cis)

cis_info <- merge(pvalue_cis, snpspos, by.x = "ID", by.y = "geneid")

pdf("Manhattan_1MB.pdf", width =15, height =7,pointsize = 15 )
cis_info %>%
  mutate(chr = as.numeric(chr)) %>%
qqman::manhattan(chr = "chr", bp = "posPruned", p = "pvalue", snp = "ID", ylim = c(0,20), cex.axis = 0.5,genomewideline = -log10(1e-08), col=rev(colorRampPalette(inferno(22))(22)),annotatePval=0.01)

dev.off()

```


## qqPlot
```{r}
pdf("qq_1MB_M.pdf", width = 12, height = 7,pointsize = 20)
plot(
  me_DNAme,
  cex = 0.5,
  pch = 19,
  xlim = NULL,
  ylim = NULL,
  main = NULL)
dev.off()
```


## Annotate cis-meQTL

```{r}
 # package for getting snp annotation
sigCis <- me_DNAme$cis$eqtls %>%
  dplyr::filter(FDR < 0.05) 

## snp annotation
table1 <- getBM(attributes = c("refsnp_id", "chr_name", 
                       "allele", "minor_allele","chrom_start","chrom_end", "ensembl_gene_stable_id"), 
                filters = "snp_filter", 
                values = sigCis$snps, 
                mart = grch37_snp) %>% 
  as.tibble() %>% 
  unique() %>% 
  dplyr::filter(!grepl("H", .$chr_name))

table2 <- getBM(attributes = c("ensembl_gene_id", "external_gene_name","description"),
                filters = "ensembl_gene_id", 
                values =  table1$ensembl_gene_stable_id, 
                mart = grch37_gene)

Cis_snp_ann <- inner_join(table1,table2, by = c("ensembl_gene_stable_id"="ensembl_gene_id"), all.x=T) %>% 
  as.data.frame()
ann_CisCpG <- ann.Epic[match(sigCis$gene, ann.Epic$Name),] %>% as.data.frame()
rownames(ann_CisCpG) <- NULL

ann_Cisme <- merge(merge(Cis_snp_ann,sigCis, by.x ="refsnp_id", by.y="snps",no.dups=TRUE), unique(ann_CisCpG), by.x = "gene", by.y = "Name") %>% 
  as.data.frame() %>%  dplyr::select(snps=refsnp_id,chr_name, 
                       allele, minor_allele,chrom_start,chrom_end, external_gene_name, description,
                       CpG = gene,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,statistic, pvalue, FDR, beta) %>% unique()
  
ann_Cisme <- ann_Cisme[order(ann_Cisme$pvalue),]
ann_Cisme10 <- ann_Cisme[1:10,] %>% dplyr::select(snps,chr=chr_name,pos=chrom_start,external_gene=external_gene_name,CpG,pos,RefGene=UCSC_RefGene_Name,statistic,pvalue,FDR,beta)
write_csv(ann_Cisme10,"annotation_cis_10.csv")
```



## Annotate trans-meQTL

```{r}
sigTrans <- me_DNAme$trans$eqtls %>%
  dplyr::filter(FDR < 0.05)

table3 <- getBM(attributes = c("refsnp_id", "chr_name", 
                       "allele", "minor_allele","chrom_start","chrom_end", "ensembl_gene_stable_id"), 
                filters = "snp_filter", 
                values = sigTrans$snps, 
                mart = grch37_snp) %>% as.tibble() %>% 
  unique() %>% 
dplyr::filter(!grepl("H", .$chr_name))

table4 <- getBM(attributes = c("ensembl_gene_id","external_gene_name","description"),
                filters = "ensembl_gene_id", 
                values =  table3$ensembl_gene_stable_id, 
                mart = grch37_gene)

Trans_snp_ann <- inner_join(table3,table4, by = c("ensembl_gene_stable_id"="ensembl_gene_id"), all.x=T) %>% 
  as.data.frame()
ann_TransCpG <- ann.Epic[match(sigTrans$gene, ann.Epic$Name),] %>% as.data.frame()
rownames(ann_TransCpG) <- NULL

ann_Transme <- merge(merge(Trans_snp_ann,sigTrans, by.x ="refsnp_id", by.y="snps",no.dups=TRUE), unique(ann_TransCpG), by.x = "gene", by.y = "Name") %>% 
  as.data.frame() %>%  dplyr::select(snps=refsnp_id,chr_name, 
                       allele, minor_allele,chrom_start,chrom_end, external_gene_name, description,
                       CpG = gene,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,statistic, pvalue, FDR, beta) %>% unique()
  
  
#methylation annotation
ann_Transme <- ann_Transme[order(ann_Transme$pvalue),]
ann_Transme10 <- ann_Transme[1:10,] %>% dplyr::select(snps,chr=chr_name,pos=chrom_start,external_gene=external_gene_name,CpG,pos,RefGene=UCSC_RefGene_Name,statistic,pvalue,FDR,beta)
write_csv(ann_Transme10,"anntation_trans_10.csv")
```


## Violin plot

Here we use boxplot to demonstrate the association between the genotype and the phenotype (methylation).

#cis-meQTL
```{r}
snp_info <- read_tsv("genotype_prunned_87_0.8_nodup.tsv") %>%
  dplyr::rename(snps = rsIDs)
methylation_info <- normDNAme_87
sigCis %>%
  dplyr::slice(1:9) %>%  # select row 1:10
  dplyr::select(snps, gene) %>% # select column snps and gene
  as_tibble() %>%
  left_join(methylation_info, by = c("gene"="ID")) %>% # find match CpG
  gather("Sample", "Methylation", starts_with("PAC")) %>% # Store sample name in Col "Sample", store correlated beta in Methylation
  filter(Sample %in% colnames(snp_info)) %>% # match with sample in snp, 960->870 rows
  split(f = .$Sample) %>% # split every PAC sample in splited tibbles
  lapply(function(x){
    left_join(x, snp_info[,c("snps", unique(x$Sample))]) %>% # allocate genotype data to every x by rsID
      set_colnames(gsub("PAC.+", "Genotype", colnames(.))) # the original colnames would be PACXXX, we change it to Genotype
  }) %>%
  bind_rows() %>% # bind the rows together
  mutate(Genotype = as.factor(Genotype)) %>%
  ggplot(aes(Genotype, Methylation)) + 
  geom_jitter(colour = "grey40", size = 0.4)+
  geom_violin(aes(fill=Genotype),width=1.4,scale = "count",trim = FALSE,alpha=0.5)+
  scale_fill_brewer(palette = brewer.pal(7, "Blues")[c(3,4,5)]) +
  geom_boxplot(width=0.1, color="black", alpha=0.2)+
  stat_summary(fun.y=median, geom = "point",fill ="orange",shape=21,size=2.5) +
  scale_x_discrete(breaks = c("0","1","2"),labels = c('RR','RA','AA'))+
labs(tag="Reference/Alternative")+
  facet_wrap(~snps, scales = "free")+
  theme_bw() +
  ggsave("top_cis.png",dpi = 320,height=7,width = 11)

```

#trans-meQTL, duplicate associations are removed
```{r}
sigTrans_nd<-sigTrans[!duplicated(sigTrans$snps),]
sigTrans %>%
  dplyr::slice(1:9) %>%  # select row 1:10
  dplyr::select(snps, gene) %>% # select column snps and gene
  as_tibble() %>%
  left_join(methylation_info, by = c("gene"="ID")) %>% # find match CpG
  gather("Sample", "Methylation", starts_with("PAC")) %>% # Store sample name in Col "Sample", store correlated beta in Methylation
  filter(Sample %in% colnames(snp_info)) %>% # match with sample in snp, 960->870 rows
  split(f = .$Sample) %>% # split every PAC sample in splited tibbles
  lapply(function(x){
    left_join(x, snp_info[,c("snps", unique(x$Sample))]) %>% # allocate genotype data to every x by rsID
      set_colnames(gsub("PAC.+", "Genotype", colnames(.))) # the original colnames would be PACXXX, we change it to Genotype
  }) %>%
  bind_rows() %>% # bind the rows together
  mutate(Genotype = as.factor(Genotype)) %>%
  ggplot(aes(Genotype, Methylation)) + 
  geom_jitter(colour = "grey40", size = 0.4)+
  geom_violin(aes(fill=Genotype),width=1.4,scale = "count",trim = FALSE,alpha=0.5)+
  scale_fill_brewer(palette = brewer.pal(7, "Blues")[c(3,4,5)]) +
  geom_boxplot(width=0.1, color="black", alpha=0.2)+
  stat_summary(fun.y=median, geom = "point",fill ="orange",shape=21,size=2.5) +
  scale_x_discrete(breaks = c("0","1","2"),labels = c('RR','RA','AA'))+
  facet_wrap(~snps,scales = "free")+
  theme_bw() +
  ggsave("violin_trans.png",dpi = 320,height=7,width = 11)

```



# Find snps affecting multiple CpG sites
```{r}
Cis_dup_snp <- sigCis[duplicated(sigCis$snps) | duplicated(sigCis$snps, fromLast = TRUE),]
write_csv(Cis_dup_snp,"dup_snps_cis.csv")
Cis_dup_cpg <- sigCis[duplicated(sigCis$gene) | duplicated(sigCis$gene, fromLast = TRUE),]
write_csv(Cis_dup_cpg,"dup_cpg_cis.csv")
Trans_dup_snp <- sigTrans[duplicated(sigTrans$snps) | duplicated(sigTrans$snps, fromLast = TRUE),]
write_csv(Trans_dup_snp,"dup_snp_trans.csv")
Trans_dup_cpg <- sigTrans[duplicated(sigTrans$gene) | duplicated(sigTrans$gene, fromLast = TRUE),]
write_csv(Trans_dup_cpg,"dup_cpg.trans.csv")
```

# Plot cis-distance vs. R-square
```{r}
cis_dis <- ann_Cisme %>% dplyr::select(snps,chr_name,chrom_start,CpG,pos,statistic,pvalue) %>% 
  as.tibble() %>% mutate(cis_dis = (pos-chrom_start)/1000)
tstat = cis_dis$statistic
r = tstat / sqrt( dfFull + tstat^2 )
R2 = r^2
cis_dis$R2 <- R2


ggplot(cis_dis,aes(x=cis_dis,y=R2)) + 
  geom_point(color=brewer.pal(7, "Greys")[7]) + 
  scale_x_continuous(limits = c(-1000,1000)) + 
  theme_bw() + 
  labs(x="SNP-CpG distance/kb",y="R-square") +
  ggsave("R-square_cis.pdf",dpi = 320,height=7,width = 11)

```

## plot Trans vs. R-square
```{r}
trans_dis <- ann_Transme %>% dplyr::select(snps,chrom_start,CpG,pos,statistic,pvalue) %>% as.tibble() %>% mutate(trans_dis = (chrom_start-start_position)/1000)
tstat = trans_dis$statistic
r = tstat / sqrt( dfFull + tstat^2 )
R2 = r^2
trans_dis$R2 <- R2


ggplot(trans_dis,aes(x=trans_dis,y=R2)) + geom_point() + scale_x_continuous(limits = c(-20000,20000)) + 
  theme_bw() + 
  labs(x="SNP-CpG distance/kb",y="R-square") +
  ggsave("R-square_trans.pdf",dpi = 320,height=7,width = 11)

```

## cis-DNA and snp proportion
```{r}
cis_DNAme_percentage <- table(as.character(sigCis$gene)) %>% 
  as.data.frame() %>% 
  mutate(perc=Freq/sum(Freq))

cis_DNAme_percentage %>% 
  group_by(Freq) %>% 
  summarise(perc = sum(perc)) %>% 
  ggplot(aes(x= Freq,y= perc*100)) + 
  geom_bar(stat="identity",fill="skyblue2") + 
  geom_text(aes(label = scales::percent(perc)),vjust=.5) + 
  labs(y = "Proportion", x="meQTL SNP per CpG site") +
  theme_minimal() +theme(axis.text=element_text(size=15),
        axis.title=element_text(size=20,face="bold"))+
  ggsave("cis_DNAme_percentage.pdf",dpi = 320,height=7,width = 11)



cis_SNP_percentage <- table(as.character(sigCis$snps)) %>% 
  as.data.frame() %>% 
  mutate(perc=Freq/sum(Freq))
cis_SNP_percentage %>% 
  group_by(Freq) %>% 
  summarise(perc = sum(perc)) %>% 
  ggplot(aes(x= Freq,y= perc*100)) + 
  geom_bar(stat="identity",fill="skyblue2") +
  geom_text(aes(label = scales::percent(perc)),vjust=.5) + 
  labs(y = "Proportion", x="CpG site per meQTL SNP") +
  theme_minimal() + 
  theme(axis.text=element_text(size=15),
        axis.title=element_text(size=20,face="bold"))+
  ggsave("cis_SNP_percentage.pdf",dpi = 320,height=7,width = 11)


median(cis_dis$cis_dis)

```

## Trans-DNA and snp proportion
```{r}
trans_DNAme_percentage <- table(as.character(sigTrans$gene)) %>% 
  as.data.frame() %>% 
  mutate(perc=Freq/sum(Freq))

trans_DNAme_percentage %>% 
  group_by(Freq) %>% 
  summarise(perc = sum(perc)) %>% 
  ggplot(aes(x= Freq,y= perc*100)) + 
  geom_bar(stat="identity",fill= brewer.pal(7, "RdYlGn")[2]) + 
  geom_text(aes(label = scales::percent(perc)),vjust=.5) + 
  labs(y = "Proportion", x="meQTL SNP per CpG site") +
  theme_minimal() +theme(axis.text=element_text(size=15),
        axis.title=element_text(size=20,face="bold"))+
  ggsave("trans_DNAme_percentage.pdf",dpi = 320,height=7,width = 11)



trans_SNP_percentage <- table(as.character(sigTrans$snps)) %>% 
  as.data.frame() %>% 
  mutate(perc=Freq/sum(Freq))

trans_SNP_percentage %>% 
  group_by(Freq) %>% 
  summarise(perc = sum(perc)) %>% 
  ggplot(aes(x= Freq,y= perc*100)) + 
  geom_bar(stat="identity",fill= brewer.pal(7, "RdYlGn")[2]) + 
  geom_text(aes(label = scales::percent(perc)),vjust=.5) + 
  labs(y = "Proportion", x="CpG site per meQTL SNP") +
  theme_minimal() + 
  theme(axis.text=element_text(size=15),
        axis.title=element_text(size=20,face="bold"))+
  ggsave("trans_SNP_percentage.pdf",dpi = 320,height=7,width = 11)

```

## SNPs in GWAS catalog
```{r}
GWAS_cata <-read_delim("gwas_catalog_v1.0-associations_e95_r2019-03-22.tsv", 
    "\t", escape_double = FALSE, trim_ws = TRUE) 
Gwas_cis <- GWAS_cata %>% dplyr::filter(GWAS_cata$SNPS %in% sigCis$snps)
Gwas_trans <- GWAS_cata %>% dplyr::filter(GWAS_cata$SNPS %in% sigTrans$snps)
Gwas_cis_unique <- Gwas_cis[!duplicated(Gwas_cis$SNPS),] 
Traits <- table(as.character(Gwas_cis_unique$`DISEASE/TRAIT`)) %>% as.data.frame()
Traits<- Traits[order(Traits$Freq,decreasing =TRUE),] 
png("GwasTrait.png", w=2300, h=1500, pointsize=100, res = 200)
 Traits %>% 
  dplyr::slice(1:15) %>% 
  ggplot(aes(x=Var1,y=Freq)) + 
  geom_bar(stat="identity",fill="#336666") +
  coord_flip() +
  labs(y="Count",x="Trait/Disease")+
  theme(legend.position = "top",panel.grid.major =element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(),axis.line = element_line(colour = "black")) +
   ggsave("GwasTrait.pdf",dpi = 320,height=7,width = 11)

Gwas_trans_unique <- Gwas_trans[!duplicated(Gwas_trans$SNPS),] 
Traits_trans <- table(as.character(Gwas_trans$`DISEASE/TRAIT`)) %>% as.data.frame()

```

## Mehylation sites location
```{r}
cis_me_group <- table(as.character(ann_Cisme$Relation_to_Island)) %>% 
  as.data.frame()


trans_me_group <- table(as.character(ann_Transme$Relation_to_Island)) %>% 
  as.data.frame()

Overall <- normDNAme_87 %>% inner_join(as.data.frame(ann.Epic),by=c("ID"="Name")) 
Overall <-table(as.character(Overall$Relation_to_Island)) %>% as.data.frame()



cis_me_group <- cis_me_group %>%  mutate(perc=Freq/sum(Freq)) %>% group_by(Var1,Freq) %>% 
  summarise(perc = sum(perc)) 
Overall <- Overall %>% mutate(perc=Freq/sum(Freq)) %>% group_by(Var1,Freq) %>% 
  summarise(perc = sum(perc)) 
trans_me_group <- trans_me_group %>% mutate(perc=Freq/sum(Freq)) %>% group_by(Var1,Freq) %>% 
  summarise(perc = sum(perc))
cis_me_group$df <- 1
trans_me_group$df <-2
Overall$df <- 3
df <- rbind(cis_me_group, trans_me_group,Overall)
df <- melt(df, id.vars=c("Var1", "perc", "df"))

ggplot(df,aes(x=Var1,y=perc*100,fill=as.factor(df)))+
  geom_bar(stat = "identity",position = 'dodge') +
  scale_fill_manual(values = c("1" = "#e97f02", "2" = "#f8ca00", "3" = "#77AAAD"),labels = c("cis-meQTL CpG","trans-meQTL CpG","EPIC array")) +
  geom_text(aes(label = scales::percent(perc)),position = position_dodge(width = .9))+
  labs(x=NULL,y="Proportion") + theme_bw()+
  theme(legend.title =element_blank(),axis.title.y = element_text(size=16),axis.text=element_text(size=16),legend.text = element_text(size = 20)) +
   ggsave("Methylation_location.pdf",dpi = 320,height=7,width = 11)

```



## Hit cis-meSNP with public regulatory database
#cis-meQTL
```{r}
cis_snp_gr <- sigCis %>%
  getBM(attributes = c("refsnp_id", "chr_name","chrom_start","chrom_end","chrom_strand"), 
                  filters = "snp_filter", 
                  values = sigCis$snps, 
                  mart = grch37_snp, uniqueRows = TRUE) %>% as.data.frame() %>% 
  unique() %>%  
  dplyr::filter(!grepl("H", .$chr_name))

cis_snp_gr$chr_name <- sub("^", "chr", cis_snp_gr$chr_name) 
cis_snp_gr <- as_granges(dplyr::distinct(cis_snp_gr,refsnp_id,.keep_all=TRUE),seqnames =chr_name,start=chrom_start,end=chrom_end,strand = chrom_strand)
regulatory_elements <- plyranges::read_bed("E091_15_coreMarks_dense.bed")
enhancer_humanAll <- read_bed("human_permissive_enhancers_phase_1_and_2.bed")
enhancer_humanAll <- read_bed("enhancer_data_at_2019-01-24_06-14-09.bed")
Promoter_humanAll <- readed("promoter_data_at_2019-01-24_08-21-25.bed")
cis_cpg_gr<- as_granges(dplyr::distinct(ann_CisCpG,Name,.keep_all=TRUE),seqnames=chr,start=pos,end=pos+1)
allSNPPro <- find_overlaps(cis_snp_gr,Promoter_humanAll,suffix = c("ranges", "ranges")) %>% 
  as.data.frame() 
allSNPEhn <-find_overlaps(cis_snp_gr,enhancer_humanAll,suffix = c("ranges", "ranges")) %>% 
  as.data.frame()
allSNPReg <-find_overlaps(cis_snp_gr,regulatory_elements,suffix = c("ranges", "ranges")) %>% 
  as.data.frame() 
allCPGPro <-find_overlaps(cis_cpg_gr,Promoter_humanAll,suffix = c("ranges", "ranges")) %>% as.data.frame() 
allCPGEhn <-find_overlaps(cis_cpg_gr,enhancer_humanAll,suffix = c("ranges", "ranges")) %>% as.data.frame() 
allCPGReg <-find_overlaps(cis_cpg_gr,regulatory_elements,suffix = c("ranges", "ranges")) %>% as.data.frame() 
allQTLReg <- sigCis %>% 
  dplyr::select(snps,gene) %>% as_tibble() %>% 
  inner_join(allCPGReg,by=c("gene"="Name")) %>% 
  dplyr::distinct() %>% 
  dplyr::select(snps,gene,CpGname=name) %>% 
  inner_join(allSNPReg,by=c("snps"="refsnp_id")) %>% 
  dplyr::distinct() %>% dplyr::select(snps,Cpg=gene,SNPname=name,CpGname)
common <- allQTLReg[allQTLReg$SNPname == allQTLReg$CpGname,]

# Plot
allSNPReg <- allSNPReg[str_order(allSNPReg$name,numberic=TRUE),] %>% as.tibble

all_SNP_Reg_percentage <- table(as.character(allSNPReg$name)) %>% 
  as.data.frame() %>% 
  mutate(perc=Freq/sum(Freq))


ChromHMM_color <- allSNPReg %>% 
  dplyr::select(name,itemRgb) %>% 
  dplyr::distinct()

ChromHMM_color_1 <- ChromHMM_color$name %>% 
    str_sort(numeric = T) %>% 
     str_remove(.,"[0-9]*_") %>% 
  as_tibble() %>%
  magrittr::set_colnames("name") %>% 
  left_join(ChromHMM_color %>% 
              mutate(name = str_remove(name,"[0-9]*_"))) %>%   
  dplyr::filter(!startsWith(name,"Q")) %>% 
  dplyr::distinct(name,itemRgb) %>% 
  column_to_rownames(var="name")

allSNPReg %>%  
  right_join(all_SNP_Reg_percentage,by=c("name"="Var1")) %>%   
  as.data.frame() %>% 
  dplyr::select(name,Freq) %>% 
  dplyr::filter(!startsWith(name,"15")) %>%
  mutate(name=gsub(".*_","",name)) %>% 
  mutate(name = factor(name, level = rev(rownames(ChromHMM_color_1)))) %>%
  ggplot(aes(name,fill=name)) + 
  geom_bar(aes(y=..count../sum(..count..))) +
  scale_fill_manual(values=rev(ChromHMM_color_1$itemRgb)) +
  scale_y_continuous(labels=percent_format())+
  coord_flip()+
  theme_bw()+
  theme(legend.position = "none",axis.title = element_blank(), axis.text=element_text(size=15)) +
  ggsave("Cis_HMM_snp.pdf",dpi = 320,height=7,width = 11)



## CPG
all_CPG_Reg_percentage <- table(as.character(allCPGReg$name)) %>% 
  as.data.frame() %>% 
  mutate(perc=Freq/sum(Freq))

ChromHMM_color_2 <- ChromHMM_color[ChromHMM_color$name %in% allCPGReg$name,]
ChromHMM_color_2 <- ChromHMM_color_2$name %>% 
  str_sort(numeric = T) %>% 
     str_remove(.,"[0-9]*_") %>% 
  as_tibble() %>%
  magrittr::set_colnames("name") %>% 
  left_join(ChromHMM_color %>% 
              mutate(name = str_remove(name,"[0-9]*_"))) %>%   
  dplyr::filter(!startsWith(name,"Q"))  %>% 
  dplyr::distinct(name,itemRgb) %>% 
  column_to_rownames(var="name")


allCPGReg %>%  
  right_join(all_CPG_Reg_percentage,by=c("name"="Var1")) %>%  
  as.data.frame() %>% 
  dplyr::select(name,Freq) %>% 
  dplyr::filter(!startsWith(name,"15")) %>% 
  mutate(name=gsub(".*_","",name)) %>%
  mutate(name = factor(name, level = rev(rownames(ChromHMM_color_2)))) %>%
  na.omit() %>% 
  ggplot(aes(name,fill=name)) + 
   geom_bar(aes(y=..count../sum(..count..))) +
  scale_fill_manual(values=rev(ChromHMM_color_2$itemRgb))+
  scale_y_continuous(labels=percent_format())+
  coord_flip() +
  theme_bw()+
  theme(legend.position = "none",axis.title = element_blank(), axis.text=element_text(size=15)) +
   ggsave("Cis_HMM_CPG.pdf",dpi = 320,height=7,width = 11)

```


#trans-meQTL
```{r}
trans_snp_gr <- sigTrans %>%
  getBM(attributes = c("refsnp_id", "chr_name","chrom_start","chrom_end","chrom_strand"), 
                  filters = "snp_filter", 
                  values = sigTrans$snps, 
                  mart = grch37_snp, uniqueRows = TRUE) %>% as.data.frame() %>% 
  unique() %>%  dplyr::filter(!grepl("H", .$chr_name))

trans_snp_gr$chr_name <- sub("^", "chr", trans_snp_gr$chr_name) 
trans_snp_gr <- as_granges(dplyr::distinct(trans_snp_gr,refsnp_id,.keep_all=TRUE),seqnames =chr_name,start=chrom_start,end=chrom_end,strand = chrom_strand)
regulatory_elements <- read_bed("E091_15_coreMarks_dense.bed")
#enhancer_humanAll <- readBed("human_permissive_enhancers_phase_1_and_2.bed")
trans_cpg_gr<- as_granges(dplyr::distinct(ann_TransCpG,Name,.keep_all=TRUE),seqnames=chr,start=pos,end=pos+1)
Trans_allSNPReg <-find_overlaps(trans_snp_gr,regulatory_elements,suffix = c("ranges", "ranges")) %>% 
  as.data.frame() 
Trans_allSNPPro <- find_overlaps(trans_snp_gr,Promoter_humanAll,suffix = c("ranges", "ranges")) %>% 
  as.data.frame() 
Trans_allSNPEhn <-find_overlaps(trans_snp_gr,enhancer_humanAll,suffix = c("ranges", "ranges")) %>% 
  as.data.frame() 
Trans_allCPGPro <-find_overlaps(trans_snp_gr,Promoter_humanAll,suffix = c("ranges", "ranges")) %>% as.data.frame()
Trans_allCPGEhn <-find_overlaps(trans_snp_gr,enhancer_humanAll,suffix = c("ranges", "ranges")) %>% as.data.frame()
Trans_allCPGReg <-find_overlaps(trans_cpg_gr,regulatory_elements,suffix = c("ranges", "ranges")) %>% as.data.frame() 
Trans_allQTLReg <- sigTrans %>% 
  dplyr::select(snps,gene) %>% as_tibble() %>% 
  inner_join(Trans_allCPGReg,by=c("gene"="Name")) %>% 
  dplyr::distinct() %>% 
  dplyr::select(snps,gene,CpGname=name)   %>% 
  inner_join(Trans_allSNPReg,by=c("snps"="refsnp_id")) %>% 
  dplyr::distinct() %>% dplyr::select(snps,Cpg=gene,SNPname=name,CpGname)
common_trans <- Trans_allQTLReg[Trans_allQTLReg$SNPname == Trans_allQTLReg$CpGname,]


## Plot

Trans_allSNPReg <- Trans_allSNPReg[order(Trans_allSNPReg$name),]
Trans_all_SNP_Reg_percentage <- table(as.character(Trans_allSNPReg$name)) %>% 
  as.data.frame() %>% 
  mutate(perc=Freq/sum(Freq))

png("trans_HMM_snp.png", w=2300, h=1200, pointsize=100, res = 200)
ChromHMM_color_3<-ChromHMM_color[ChromHMM_color$name %in% Trans_allSNPReg$name,]
ChromHMM_color_3 <- ChromHMM_color_3$name %>% 
     str_sort(numeric = T) %>% 
     str_remove(.,"[0-9]*_") %>% 
  as_tibble() %>%
  magrittr::set_colnames("name") %>% 
  left_join(ChromHMM_color %>% 
              mutate(name = str_remove(name,"[0-9]*_"))) %>% 
  dplyr::filter(!startsWith(name,"Q"))  %>% 
  dplyr::distinct(name,itemRgb) %>% 
  column_to_rownames(var="name") 

Trans_allSNPReg %>%  
  right_join(Trans_all_SNP_Reg_percentage,by=c("name"="Var1")) %>%  
  as.data.frame() %>% 
  dplyr::select(name,Freq) %>% 
  dplyr::filter(!startsWith(name,"15")) %>% 
  mutate(name=gsub(".*_","",name)) %>% 
  mutate(name = factor(name, level = rev(rownames(ChromHMM_color_3)))) %>%
  na.omit() %>% 
  ggplot(aes(name,fill=name)) + 
   geom_bar(aes(y=..count../sum(..count..))) +
  scale_fill_manual(values=rev(ChromHMM_color_3$itemRgb))+
  scale_y_continuous(labels=percent_format())+
  labs(y=gsub("_","",ChromHMM_color$name)) +
  coord_flip() +
  theme_bw()+
  theme(legend.position = "none",axis.title = element_blank(), axis.text=element_text(size=15))+
  ggsave("trans_HMM_snp.pdf",dpi = 320,height=7,width = 11)
dev.off()

## CPG
Trans_all_CPG_Reg_percentage <- table(as.character(Trans_allCPGReg$name)) %>% 
  as.data.frame() %>% 
  mutate(perc=Freq/sum(Freq))


ChromHMM_color_4 <- ChromHMM_color[ChromHMM_color$name %in% Trans_allCPGReg$name,]
ChromHMM_color_4 <- ChromHMM_color_4$name %>% 
    str_sort(numeric = T) %>% 
     str_remove(.,"[0-9]*_") %>% 
  as_tibble() %>%
  magrittr::set_colnames("name") %>% 
  left_join(ChromHMM_color %>% 
              mutate(name = str_remove(name,"[0-9]*_"))) %>%   
  dplyr::filter(!startsWith(name,"Q")) %>%
  dplyr::distinct(name,itemRgb) %>% 
  column_to_rownames(var="name")

Trans_allCPGReg %>%  
  right_join(Trans_all_CPG_Reg_percentage,by=c("name"="Var1")) %>%   
  as.data.frame() %>% 
  dplyr::select(name,Freq) %>% 
  dplyr::filter(!startsWith(name,"15")) %>% 
  mutate(name=gsub(".*_","",name)) %>% 
  mutate(name = factor(name, level = rev(rownames(ChromHMM_color_4)))) %>%
  na.omit() %>% 
  ggplot(aes(name,fill=name)) + 
   geom_bar(aes(y=..count../sum(..count..))) +
  scale_fill_manual(values=rev(ChromHMM_color_4$itemRgb))+
  scale_y_continuous(labels=percent_format())+
  coord_flip() +
  theme_bw()+
  theme(legend.position = "none",axis.title = element_blank(), axis.text=element_text(size=15)) +
  ggsave("trans_HMM_CPG.pdf",dpi = 320,height=7,width = 11)


```


## GTEx
```{r}
setwd("GTEx_Analysis_v7_eQTL")
filelist = list.files(pattern = ".*.txt")
datalist =lapply(filelist, FUN=read.delim, header=TRUE)
left_join(as.data.frame(datalist[1]),sigCis,by=c("rs_id_dbSNP147_GRCh37p13"="snps"))

Ovary <- inner_join(sigCis,as.data.frame(datalist[34]),by=c("snps"="rs_id_dbSNP147_GRCh37p13"))
#35Pancreas-41
dim(inner_join(sigCis,as.data.frame(datalist[35]),by=c("snps"="rs_id_dbSNP147_GRCh37p13")))
#Pituitary
#37Prostate-16
Prostate<-inner_join(sigCis,as.data.frame(datalist[37]),by=c("snps"="rs_id_dbSNP147_GRCh37p13"))
#39-40 skin
#Small_Intestine_Terminal_Ileum
#Spleen
#Stomach
#43Testis-23
Testis <- inner_join(sigCis,as.data.frame(datalist[43]),by=c("snps"="rs_id_dbSNP147_GRCh37p13"))
#Thyroid
#45Uterus-14
Uterus <- inner_join(sigCis,as.data.frame(datalist[45]),by=c("snps"="rs_id_dbSNP147_GRCh37p13"))
#46Vagina-15
Vagina<-inner_join(sigCis,as.data.frame(datalist[46]),by=c("snps"="rs_id_dbSNP147_GRCh37p13"))
#Whole_Blood

G <- rbind(Ovary,Prostate,Testis,Uterus,Vagina)
setwd("~/R/Weixiong_Project/data/Weixiong_DNAme_files")
```

